torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -711,22 +740,100 @@ namespace prove {
 // - x can be either zero or non-zero, it is just a symbolic number that depends
 // - x is zero
 
-bool isNonZero(Val* value) {
+bool isPositive(Val* value, const VarInfo& var_info);
+
+bool isNonNegative(Val* value, const VarInfo& var_info) {
   value = foldConstants(value);
-  if (value->getInt().has_value() && *value->getInt() != 0) {
+  if (value->getInt().has_value() && *value->getInt() >= 0) {
     return true;
   }
-  if (value->getDouble().has_value() && *value->getDouble() != 0.0) {
+  if (value->getDouble().has_value() && *value->getDouble() >= 0.0) {
+    return true;
+  }
+  if (auto ns = dynamic_cast<NamedScalar*>(value)) {
+    if (ns->getParallelDim().has_value() ||
+        ns->getParallelIndex().has_value() || ns->isTensorSize() ||
+        ns->isTensorStride()) {
+      return true;
+    }
+  }
+  value = maybeUnwrapMagicZero(value);
+  if (var_info.has(value)) {
+    return isNonNegative(var_info.info(value).start, var_info) &&
+        isNonNegative(var_info.info(value).step, var_info);
+  }
+  if (auto fop = dynamic_cast<FOp*>(value->definition())) {
+    auto op = fop->getOpType();
+    if (op == BinaryOpType::Add || op == BinaryOpType::Mul) {
+      for (auto inp : fop->inputs()) {
+        if (!isNonNegative(inp, var_info)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  } else if (auto bop = dynamic_cast<BinaryOp*>(value->definition())) {
+    auto op = bop->getBinaryOpType();
+    if (op == BinaryOpType::Mod || op == BinaryOpType::Div ||
```

Comment:
Why are only these ops considered here? Why not, e.g., `BinaryOpType::Add`?

Reply:
Because `Add` and `Mul` are already flattened as `FlattenedAssocCommOp`, so we will not see `BinaryOp` with `Add` and `Mul` here.

Reply:
Where is the fattening done? Does this function assume the `value` parameter is flattened?

Reply:
Flatten is done before running any other passes:
https://github.com/csarofeen/pytorch/blob/2ca39e3f392aa17118a9818ff30ee650a2fcfd96/third_party/nvfuser/csrc/expr_simplifier.cpp#L1183
So all other passes assumes the input `value` already flattened.

Reply:
OK, thanks. Could you mention that as a comment? Probably obvious to you, but it would have been definitely a helpful comment for me.

Reply:
> OK, thanks. Could you mention that as a comment? Probably obvious to you, but it would have been definitely a helpful comment for me.

Oh, sorry, missed this comment. Will add to https://github.com/csarofeen/pytorch/pull/2275

Reply:
Added in https://github.com/csarofeen/pytorch/pull/2275/commits/dbe16cd78600da03f0fbe035516eadd3a5e845ef